src/query/execute_statement.c

```diff
@@ -14504,6 +14560,122 @@ do_prepare_session_statement (PARSER_CONTEXT * parser, PT_NODE * statement)
   return NO_ERROR;
 }
 
+/*
+ * do_prepare_cte () - prepare CTE statements in WITH clause for query cache
+ * return : Error code
+ * parser (in)	  : parser context
+ * statement (in) : statement to prepare
+ */
+static int
+do_prepare_cte (PARSER_CONTEXT * parser, PT_NODE * statement)
+{
+  int error = NO_ERROR;
+  PT_NODE *cte_statement;
+  PT_NODE *cte_def_list;
+  PARSER_CONTEXT cte_context;
+  int var_count;
+
+  for (cte_def_list = statement; cte_def_list; cte_def_list = cte_def_list->next)
+    {
+      if (cte_def_list->info.cte.non_recursive_part->info.query.hint & PT_HINT_QUERY_CACHE)
+	{
+	  cte_statement = cte_def_list->info.cte.non_recursive_part;
+
+	  cte_context = *parser;
+	  cte_context.dbval_cnt = 0;
+	  cte_context.host_var_count = cte_context.auto_param_count = 0;
+
+	  var_count = parser->host_var_count + parser->auto_param_count;
+
+	  cte_context.host_variables = (DB_VALUE *) parser_alloc (parser, var_count * sizeof (DB_VALUE));
+	  if (cte_context.host_variables == NULL)
+	    {
+	      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, var_count * sizeof (DB_VALUE));
+	      return ER_OUT_OF_VIRTUAL_MEMORY;
+	    }
+
+	  cte_context.cte_host_var_index = (int *) parser_alloc (parser, var_count * sizeof (int));
+	  if (cte_context.cte_host_var_index == NULL)
+	    {
+	      er_set (ER_ERROR_SEVERITY, ARG_FILE_LINE, ER_OUT_OF_VIRTUAL_MEMORY, 1, var_count * sizeof (int));
+	      return ER_OUT_OF_VIRTUAL_MEMORY;
+	    }
+
+	  cte_context.host_var_count = 0;
+
+	  parser_walk_tree (&cte_context, cte_statement, pt_cte_host_vars_index, parser, NULL, NULL);
+
+	  cte_statement->cte_host_var_count = cte_context.host_var_count;
+	  cte_statement->cte_host_var_index = cte_context.cte_host_var_index;
+
+	  error = do_prepare_select (&cte_context, cte_statement);
+	  if (error != NO_ERROR)
+	    {
+	      break;
+	    }
+	}
+    }
+
+  return error;
+}
+
+/*
+ * do_execute_cte () - execute CTE statements in WITH clause for query cache
+ * return : Error code
+ * parser (in)	  : parser context
+ * statement (in) : statement to execute
+ * query_flag     : query flag for execution
+ */
+static int
+do_execute_cte (PARSER_CONTEXT * parser, PT_NODE * statement, int query_flag)
+{
+  PT_NODE *stmt;
+  PT_NODE *cte_list;
+  QUERY_ID query_id;
+  QFILE_LIST_ID *list_id;
+  DB_VALUE *host_variables;
+  CACHE_TIME clt_cache_time;
+  int err, i, flag = query_flag & ~EXECUTE_QUERY_WITH_COMMIT;
+
+  assert (statement && statement->info.query.with);
+
+  cte_list = statement->info.query.with->info.with_clause.cte_definition_list;
+
+  CACHE_TIME_RESET (&clt_cache_time);
+
+  while (cte_list)
+    {
+      stmt = cte_list->info.cte.non_recursive_part;
+      if (stmt && (stmt->info.query.hint & PT_HINT_QUERY_CACHE))
+	{
+	  host_variables = (DB_VALUE *) parser_alloc (parser, sizeof (DB_VALUE) * stmt->cte_host_var_count);
```

Comment:
In parser_support.c, host_variable array is allocated by malloc/realloc.

Reply:
This is related to query parsing, so I keep the logic using parser_alloc.

Reply:
It is about lifecycle of an allocated memory by parser_alloc (). When and where the `cte_context.host_variables` is freed?

Reply:
I couldn't find a free function corresponding to parser_alloc(). Maybe the allocated memory will be freed depends on other object's lifecycle.

Reply:
The parser alloced memory is free when the parser end by parser_free_parser function.